---
title: "tagMeppr"
author: "Robin H. van der Weide"
date: "`r Sys.Date()`"
output: rmarkdown::pdf_document
bibliography: tagmentation-based-mapping-(tagmap)-of-mobile-dna-genomic-insertion-sites.bib
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

TRAVIS <- !identical(tolower(Sys.getenv("TRAVIS")), "true")
knitr::opts_chunk$set(purl = TRAVIS)

```


```{r logo, eval = TRAVIS, out.width = "25%", fig.align = 'center', echo = F, fig.path="vignettes"}
knitr::include_graphics("logo_tagMeppr.pdf")
```

Hi there! Welcome! 

TagMap is a very useful method for transposon mapping [@Stern037762], enabling 
researchers to map the insertion sites with ease and generate long sequencing 
reads. However, there is little to none automatisation and downstream analysis 
software available for these reads. TagMeppr is an easy to use, memory 
efficient fastq-to-figure package written in R.

## Usage
The basic usage of tagMapper revolves around three clear steps:

1. index: a tagMapper-index is made once for a specific genome and protocol (e.g. hg19 and PigyBac).
2. align: a tagMapperSample-object is made and aligned to the index
3. analyse: determine and plot highly likely integraton-sites

```{r lib, eval = TRAVIS}
library(tagMeppr)
```

## tagMepprIndex 

Remember: you only have to make a tagMepprIndex once per genome and protocol! 
This means that you can use it for however many samples for that protocol and 
organism (e.g. PiggyBac in human cells) you may have.

```{r makeIndex, eval = F, message=FALSE}
library("BSgenome.Hsapiens.UCSC.hg19")

reference_hg19_PB = makeIndex(indexPath = '../premadeReferences/', 
                              bsgenome = BSgenome.Hsapiens.UCSC.hg19, 
                              ITR = 'PiggyBac', 
                              targetInsertionSite = 'TTAA', verbose = T)
```

Loading a tagMepprIndex is easy with loadIndex().

```{r makeIndex2, eval = TRAVIS, echo = F, results='hide', message=FALSE}
library("BSgenome.Hsapiens.UCSC.hg19",quietly = T)
```

```{r loadIndex, eval = TRAVIS, cache=F, cache.lazy=F}
reference_hg19_PB = loadIndex('BSgenome.Hsapiens.UCSC.hg19_PiggyBac_tagMeppRindex.fa.gz')
```

You can use print() on this tagMepprIndex, which gives you a lot of information: 
```{r, eval = TRAVIS, cache=T, cache.lazy=F}
print(reference_hg19_PB)
```

## tagMepprSample
The core object of tagMeppr is the `tagMepprSample`. Everything, from the paths of 
the fastq's to the final results, are stored in this object. To kick off the 
analysis, such an object should be made for every sample by using `newTagMeppr()`. 
In this function, the paths of the four respective fastq's, the name of the 
sample and the transposon are asked.

```{r newTagMeppr, eval = TRAVIS, cache=F, cache.lazy=F}
C91 = newTagMeppr(F1 = 'clone91_FWD_R1.fq.gz',
                 F2 = 'clone91_FWD_R2.fq.gz',
                 R1 = 'clone91_REV_R1.fq.gz',
                 R2 = 'clone91_REV_R2.fq.gz',
                 name = "clone9 (minimised)",
                 protocol = 'PiggyBac')
```

When you use the `print()` method, a summary will be given about the sample and 
its progression trough the tagMeppr-pipeline.
```{r, eval = TRAVIS, cache=F, cache.lazy=F}
print(C91)
```


### primer-checks 
After this, it is a good idea to check whether your primer-design is one that 
tagMeppr expects: reverse- and forward-primers on different ITRs. Also, it 
checks if the reverse is found near the start of the transposon. This is to be 
able to accurately denote the orientation of the transposon. Normally, we 
expect the reverse-primer to be found at the start of the ITR-sequence. When 
running `checkPrimer()`, it will automatically update the tagMepprSample-object 
with the `rev5_fwd3` tag. If it appears to be to other way around, the sample-
object will have a `rev5_fwd3=T` flag.

It is highly recommended to do this!
```{r primerCheck, eval = TRAVIS}
fwdPrimer = "CGTCAATTTTACGCAGACTATC"
revPrimer = "GTACGTCACAATATGATTATCTTTCTAG"

checkPrimer(fwdPrimer = fwdPrimer, 
            revPrimer = revPrimer, 
            exp = C91, 
            ITR = 'PiggyBac')
```

```{r, eval = TRAVIS, cache=F, cache.lazy=F}
print(C91)
```

## Align
To align the tagMeppr-sample to the index, use the `align()`-function. This will 
call `bwa mem` and run it with optimal parameters set. If `bwa` or `samtools` 
are not installed, the function will can an error. Afterwards, the chimeric 
reads will be filtered and placed in a `GRanges` object for easy usage later 
on. Running `align()` will automatically update the `tagMeppr-sample` to 
include the parsed reads.

```{r align, eval = TRAVIS, cache=F, cache.lazy=F, message=FALSE,results='hide'}
align(exp = C91, 
      ref = reference_hg19_PB, 
      cores = 30, 
      empericalCentre = T)
```

The object will be updated, which keeps your evironment tidy. The summary will 
now also show the number of informative reads (i.e. a chimeric read to both an 
ITR and the reference) and the temporary folder of the .BAM-files.
```{r, eval = TRAVIS, cache=F, cache.lazy=F}
print(C91)
```

## Analysis

### Find insertions
The `findInsertions()` function will first find all reads that overlap a TIS, 
which in the case of PiggyBac will be "TTAA". Next it will calculate whether 
there is a bias towards one side of the TIS using a binominal test. The bias, 
denoted as $D$, $-1$ when all reads are upstream and $+1$ when all reads are 
downstream of the TIS.This is done independently for the forward and reverse 
reads:

$p_{fwd/rev} = \binom{reads_{D<0}}{reads}$

Next, we filter out TISs which have the bias on the same side of the TIS:

$sgn(D_{fwd}) \neq sgn(D_{rev})$

To calculate a "TIS-specific" p-value, we use Edgington's sum-p method, which 
is very conservative in our usage. This ensures that, when 
$p_{combined} < \alpha$, both the fwd and the rev reads are indeed biased.

$p_{combined} = \dfrac{(\sum_{i=1}^{2} p_i)^{2}}{2!}$

Afterwards, a holm-correction is done to limit the Family-Wise Error Rate (FWER).

```{r FI, eval = TRAVIS, cache=F, cache.lazy=F}
findInsertions(exp = C91, ref = reference_hg19_PB, padding = 2)
```

```{r, eval = TRAVIS, cache=F, cache.lazy=F}
print(C91)
```

```{r, eval = TRAVIS, cache=F, cache.lazy=F, fig.align='center', fig.height= 3, fig.width=3}
foundInsertions = results( C91 )

head(foundInsertions)

barplot(table(foundInsertions$strand), horiz = F,
        col = tagMepprCol(),
        main = 'found orientations')
```

### Plot

#### Single insertions
```{r PLOTsingle, eval = TRAVIS, fig.width=10, dpi=300, cache=T, cache.lazy=F, fig.align='center', out.width='50%', warning=FALSE, dev=c('pdf','png')}
plotSite(C91,site = 3)
```

```{r PLOTsingle100, eval = TRAVIS, fig.width=10, dpi=300,cache=T, cache.lazy=F, fig.align='center', out.width='50%', warning=FALSE, dev=c('pdf','png')}
plotSite(C91,site = 3, maxReads = 100)
```

#### All insertions

```{r PI, eval = TRAVIS, cache=F, fig.width=10, dpi=300,cache.lazy=F, dev=c('pdf','png'), fig.align='center', out.width='50%'}
plotInsertions(exp = C91)
```

show orientation:
```{r PIo, eval = TRAVIS, cache=F, fig.width=10, dpi=300,cache.lazy=F, dev=c('pdf','png'), fig.align='center', out.width='50%'}
plotInsertions(exp = C91, showOrientation = T)
```

##### Plot muliple samples

```{r makeList, eval = TRAVIS, echo = F, cache=F, cache.lazy=F}
C91_a = C91
C91_b = C91

C91_a$name = 'clone9 subset 1'
C91_b$name = 'clone9 subset 2'

C91_a$results = sample(C91_a$results, size = floor(length(C91_a$results)/2))
C91_b$results = sample(C91_b$results, size = floor(length(C91_b$results)/2))

tagMepprSampleList = list(C91_a, C91_b)
```


```{r PImult1, eval = TRAVIS, fig.width=10, dpi=300, dev=c('pdf','png'), fig.align='center', cache.lazy=F, out.width='50%'}
plotInsertions(tagMepprSampleList)
```

```{r PImult1o, eval = TRAVIS, fig.width=10, dpi=300,cache=F, dev=c('pdf','png'), fig.align='center', cache.lazy=F, out.width='50%'}
plotInsertions(tagMepprSampleList, showOrientation = T)
```


```{r PImult2, eval = TRAVIS, fig.width=10, dpi=300,cache=F, dev=c('pdf','png'), cache.lazy=F, fig.align='center', out.width='50%'}
plotInsertions(tagMepprSampleList, sideBySide = T)
```

## Biography
