---
title: "tagMeppr"
author: "Robin H. van der Weide"
date: "`r Sys.Date()`"
output: rmarkdown::pdf_document
bibliography: tagmentation-based-mapping-(tagmap)-of-mobile-dna-genomic-insertion-sites.bib
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE, echo=FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)

TRAVIS <- !identical(tolower(Sys.getenv("TRAVIS")), "true")
knitr::opts_chunk$set(purl = TRAVIS)

```


```{r logo, eval = TRAVIS, out.width = "25%", fig.align = 'center', echo = F, fig.path="vignettes"}
knitr::include_graphics("logo_tagMeppr.pdf")
```

Hi there! Welcome! 

TagMap is a very useful method for transposon mapping [@Stern037762], enabling 
researchers to map the insertion sites with ease and generate long sequencing 
reads. However, there is little to none automatisation and downstream analysis 
software available for these reads. TagMeppr is an easy to use, memory 
efficient fastq-to-figure package written in R.

```{r lib, eval = T}
library(tagMeppr)
```

# A hybrid reference genome 
Remember: you only have to make a tagMepprIndex once per genome and protocol! 
This means that you can use it for however many samples for that protocol and 
organism (e.g. PiggyBac in human cells) you may have.

```{r makeIndex, eval = F}
library("BSgenome.Hsapiens.UCSC.hg19")

reference_hg19_PB = makeIndex(indexPath = '../premadeReferences/', 
                              bsgenome = BSgenome.Hsapiens.UCSC.hg19, 
                              ITR = 'PiggyBac', 
                              targetInsertionSite = 'TTAA', verbose = T)
```

Loading a tagMepprIndex is easy with loadIndex().

```{r makeIndex2, eval = T, echo = F, results='hide'}
library("BSgenome.Hsapiens.UCSC.hg19")
```

```{r loadIndex, eval=T, cache=F, cache.lazy=F}
reference_hg19_PB = loadIndex('BSgenome.Hsapiens.UCSC.hg19_PiggyBac_tagMeppRindex.fa.gz')
```

You can use print() on this tagMepprIndex, which gives you a lot of information: 
```{r, eval=T, cache=T, cache.lazy=F}
print(reference_hg19_PB)
```

# The tagMepprSample-object

## Creating a new sample

```{r newTagMeppr, eval=T, cache=F, cache.lazy=F}
C91 = newTagMeppr(F1 = 'clone91_FWD_R1.fq.gz',
                 F2 = 'clone91_FWD_R2.fq.gz',
                 R1 = 'clone91_REV_R1.fq.gz',
                 R2 = 'clone91_REV_R2.fq.gz',
                 name = "clone9 (minimised)",
                 protocol = 'PiggyBac')
```

This is the print:
```{r, eval=T, cache=F, cache.lazy=F}
print(C91)
```


## checken 
Normally, we expect the reverse-primer to be found at the start of the ITR-
sequence. This can be checked with `checkPrimer()`. If it appears to be to 
other way around, the sample-object will have a `flipPrimer=T` flag.

Highly recommend this!
```{r, primerCheck}
fwdPrimer = "CGTCAATTTTACGCAGACTATC"
revPrimer = "GTACGTCACAATATGATTATCTTTCTAG"

checkPrimer(fwdPrimer = fwdPrimer, 
            revPrimer = revPrimer, 
            exp = C91, 
            ITR = 'PiggyBac')
```
```{r, eval=T, cache=F, cache.lazy=F}
print(C91)
```

# Aligning
Omg. Mapping without leaving R! Please check if you have BWA and samtools intalled.
Note no <- signs

```{r align, eval=T, cache=T, cache.lazy=F, message=FALSE,results='hide'}
align(exp = C91, 
      ref = reference_hg19_PB, 
      cores = 30, 
      empericalCentre = T)
```



The object will be updated, which keeps your evironment tidy.
```{r, eval=T, cache=F, cache.lazy=F}
print(C91)
```

# Find insertions


```{r FI, eval=T, cache=F, cache.lazy=F}
findInsertions(exp = C91, ref = reference_hg19_PB, padding = 2)
```

```{r, eval=T, cache=F, cache.lazy=F}
print(C91)
```

```{r, eval=T, cache=F, cache.lazy=F, fig.align='center', fig.height= 3, fig.width=3}
foundInsertions = results( C91 )

head(foundInsertions)

barplot(table(foundInsertions$strand), horiz = F,
        col = tagMepprCol(),
        main = 'found orientations')
```

# Plot

## Single insertions
```{r PLOTsingle, eval=T, cache=T, cache.lazy=F, fig.align='center', out.width='50%', warning=FALSE}
plotSite(C91,site = 3)
```

```{r PLOTsingle100, eval=T, cache=T, cache.lazy=F, fig.align='center', out.width='50%', warning=FALSE}
plotSite(C91,site = 3, maxReads = 100)
```

## All insertions

```{r PI, eval=T, cache=F, cache.lazy=F, fig.align='center', out.width='50%'}
plotInsertions(exp = C91)
```

show orientation:
```{r PIo, eval=T, cache=F, cache.lazy=F, fig.align='center', out.width='50%'}
plotInsertions(exp = C91, showOrientation = T)
```

## Plot muliple samples

```{r makeList, eval=T, echo = F, cache=F, cache.lazy=F}
C91_a = C91
C91_b = C91

C91_a$name = 'clone9 subset 1'
C91_b$name = 'clone9 subset 2'

C91_a$results = sample(C91_a$results, size = floor(length(C91_a$results)/2))
C91_b$results = sample(C91_b$results, size = floor(length(C91_b$results)/2))

tagMepprSampleList = list(C91_a, C91_b)
```


```{r PImult1, eval=T, cache=F, fig.align='center', cache.lazy=F, out.width='50%'}
plotInsertions(tagMepprSampleList)
```

```{r PImult1o, eval=T, cache=F, fig.align='center', cache.lazy=F, out.width='50%'}
plotInsertions(tagMepprSampleList, showOrientation = T)
```


```{r PImult2, eval=T, cache=F, cache.lazy=F, fig.align='center', out.width='50%'}
plotInsertions(tagMepprSampleList, sideBySide = T)
```

# Biography
